README.md
# @grantcodes/topography-paintlet

A css paintlet to create a nice custom topography inspired background.

## Usage

### 1. Register the paint worklet in JavaScript:

```javascript
CSS.paintWorklet.addModule(
  'https://unpkg.com/@grantcodes/topography-paintlet/dist/topography-paintlet.js'
)
```

NOTE: This example loads the worklet from unpkg.com, but you could also copy and host the file locally.

### 2. Set up the colors and corner in CSS and use the `paint` function.

```css
.topography {
  background-color: #2a4858;
  background-position: --topography-corner;
  background-size: cover;
  background-repeat: no-repeat;
  background-image: paint(grantcodes-topography);
  --topography-corner: top right;
  --topography-colors: #106e7c, #00968e, #4abd8c, #9cdf7c, #fafa6e;
}
```

### Random gotchas

- Works best with 5 colors in my opinion but it should work if you use more.
- Doesn't seem to work great when applied to the `html` or `body` elements.


CHANGELOG.md
# Changelog

## [1.0.1](https://github.com/grantcodes/topography-paintlet/compare/v1.0.0...v1.0.1) (2022-08-14)


### Bug Fixes

* **actions:** only setup node on release creation ([dfc5f98](https://github.com/grantcodes/topography-paintlet/commit/dfc5f986410c70abb8c7f4a7d5fdb0caa3713b3e))

## 1.0.0 (2022-08-14)


### Bug Fixes

* **actions:** update actions publish version ([f6db4c5](https://github.com/grantcodes/topography-paintlet/commit/f6db4c55be28eee36c37c9cf418fb40647be7393))
* add tsup dependency ([54f8427](https://github.com/grantcodes/topography-paintlet/commit/54f84278e94cb75b294b0b5652acd93407562833))


src/tests/helpers.test.ts
import test from 'ava'

import { randomInt, radians } from '../lib/helpers'

test('randomInt', t => {
  t.is(randomInt(2, 2), 2)
  const res = randomInt(20, 200)
  t.truthy(res >= 20 && res <= 200)
  t.truthy(Number.isInteger(res))
})

test('radians', t => {
  t.is(radians(30), 0.5235987756)
  t.is(radians(90), 1.570796327)
  t.is(radians(-270), -4.71238898)
})


src/tests/get-noise.test.ts
import test from 'ava'

import { getNoise } from '../lib/get-noise'

test('getNoise', t => {
  const noise1 = getNoise(0.522, 1)
  const noise2 = getNoise(0.522, 1)
  const noise3 = getNoise(0.522, 3)
  const noise4 = getNoise(1.211, 5)

  // Make sure between 0 and 1
  t.truthy(noise1 > 0 && noise1 < 1)
  t.truthy(noise2 > 0 && noise2 < 1)
  t.truthy(noise3 > 0 && noise3 < 1)
  t.truthy(noise4 > 0 && noise4 < 1)

  // Make sure it generates the same number with same input
  t.is(noise1, noise2)
})


src/tests/get-shapes.test.ts
import test from 'ava'

import { getShapes } from '../lib/get-shapes'

test('getShapes', t => {
  const shapes = getShapes(7)

  // Make sure it generates the given number of shapes
  t.is(shapes.length, 7)

  // Make sure shapes are an array of objects like {x, y}
  t.assert(Array.isArray(shapes[0]))
  const firstPoint = shapes[0][0]
  t.is(typeof firstPoint, 'object')
  t.assert('x' in firstPoint)
  t.assert('y' in firstPoint)
})


src/lib/get-shapes.ts
import { radians } from './helpers'
import { getNoise } from './get-noise'
import { SIZE, ANGLE_START, ANGLE_END, ANGLE_COUNT } from './consts'
import { ShapePoint, Shape } from '../types'

const shapes: Shape[] = []

function getShapes (count: number): Shape[] {
  if (shapes.length === count) {
    return shapes
  }

  const magnitude = SIZE / count // Bigger number = bigger rings

  // Create array of angles for the shape.
  const angles: number[] = []
  for (let ang = ANGLE_START; ang <= ANGLE_END; ang = ang + 90 / ANGLE_COUNT) {
    angles.push(ang)
  }

  // Create arrays of points for each shape for each layer.
  for (let layerI = 0; layerI < count; layerI++) {
    const reverseLayerI = count - layerI
    const shape: Shape = angles.map(
      (angle: number): ShapePoint => {
        const rad = radians(angle)
        const noiseVal = getNoise(rad, reverseLayerI)
        const newRadius = noiseVal * reverseLayerI * magnitude
        const x = Math.min(newRadius * Math.cos(rad), SIZE)
        const y = Math.min(newRadius * Math.sin(rad), SIZE)

        return { x, y }
      }
    )

    shapes.push(shape)
  }

  return shapes
}

export { getShapes }


src/lib/TopographyPaintlet.ts
import { SIZE, DEFAULT_COLORS, CSS_CUSTOM_PROPERTIES } from './consts'
import { radians } from './helpers'
import { getShapes } from './get-shapes'
import { ShapeCorner, ShapePoint } from '../types'

interface PaintSize {
  width: number
  height: number
}

class TopographyPaintlet {
  static get inputProperties (): string[] {
    return Object.values(CSS_CUSTOM_PROPERTIES)
  }

  static get contextOptions (): any {
    return { alpha: true }
  }

  paint (
    ctx: CanvasRenderingContext2D,
    size: PaintSize,
    properties: any
  ): void {
    const maxSize = Math.min(size.width, size.height)

    // Get colors from --topography-colors property.
    let colors = DEFAULT_COLORS
    const colorsString: string =
      properties.get(CSS_CUSTOM_PROPERTIES.colors)?.toString() ?? ''
    const commaRegex = /(?:(["'])(\\.|(?!\1)[^\\])*\1|\[(?:(["'])(\\.|(?!\2)[^\\])*\2|[^\]])*\]|\((?:(["'])(\\.|(?!\3)[^\\])*\3|[^)])*\)|[^,])+/g // eslint-disable-line no-useless-backreference
    const matches = colorsString.match(commaRegex)
    if (Array.isArray(matches)) {
      colors = matches.map((c: string) => c.trim()).filter((c: string) => c)
    }

    const layerCount = colors.length

    // Get positioning form --topography-corner property.
    const corner: ShapeCorner =
      properties
        .get(CSS_CUSTOM_PROPERTIES.corner)
        ?.toString()
        ?.trim() ?? 'top right'

    // Define where the shapes start from
    const zeroPoint: ShapePoint = { x: 0, y: 0 }

    // Load the shapes into the class property to cache them.
    const shapes = getShapes(layerCount)

    // Scale the image to fit the given size
    const scaleAmount = maxSize / SIZE
    ctx.scale(scaleAmount, scaleAmount)

    const canvasWidth = size.width / scaleAmount
    const canvasHeight = size.height / scaleAmount

    // Rotate to fit the correct corner.
    if (corner === 'top right') {
      ctx.rotate(radians(90))
      ctx.translate(0, 0 - canvasWidth)
    } else if (corner === 'bottom right') {
      ctx.rotate(radians(180))
      ctx.translate(0 - canvasWidth, 0 - Math.max(SIZE, canvasHeight))
    } else if (corner === 'bottom left') {
      ctx.rotate(radians(270))
      ctx.translate(0 - Math.max(SIZE, canvasHeight), 0)
    }

    // Draw a path for each layer.
    let layerI = 0
    for (const shape of shapes) {
      const color = colors[layerI]
      ctx.fillStyle = color
      // Move to start position
      ctx.moveTo(zeroPoint.x, zeroPoint.y)

      // Transform offscreen if enter property set.
      // TODO: This could be used to create an entry animation for the shapes.
      // const enter = properties.get('--topography-enter')?.toString();
      // if (enter) {
      //   const percent = parseInt(enter);
      //   if (percent) {
      //     const shapeEnterStart = (100 / layerCount) * layerI;
      //     const shapeProgress = Math.min(
      //       (percent - shapeEnterStart) * layerCount,
      //       100
      //     );
      //     const amount =
      //       shapeProgress > 0
      //         ? maxSize - (shapeProgress / 100) * maxSize
      //         : maxSize;
      //     ctx.translate(amount, 0 - amount);
      //   }
      // }

      // Start the path.
      ctx.beginPath()

      // Draw curves between the points
      let pointI = 0
      for (const point of shape) {
        const nextPoint =
          shape[pointI < shape.length - 1 ? pointI + 1 : shape.length - 1]
        const xc = (point.x + nextPoint.x) / 2
        const yc = (point.y + nextPoint.y) / 2

        ctx.quadraticCurveTo(point.x, point.y, xc, yc)
        pointI++
      }

      // Finish off the shape by going back to the start position.
      ctx.lineTo(zeroPoint.x, zeroPoint.y)
      ctx.fill()

      ctx.closePath()
      layerI++
    }
  }
}

export { TopographyPaintlet }
